feat: Enable auto registration of view models

mishael-o opened the pull request:
# Description

This change implements improvements and features to the library. They are outlined below.

## Breaking Changes

- Deprecated `AddMvvmNavigation`. Consumers should use `AddMvvm`.
- Deprecated `Loaded()` method on `IViewModelBase`. Consumers should use `OnInitializedAsync()`.
- Deprecated `MvvmValidationSummary`. Consumers should use the `ValidationSummary` summary provided by the Blazor SDK, and the new `MvvmObservableValidator` provided by the library.
- Dropped support for .NET 7 as it has reached EOL.

## Features

- Added auto registration and discovery of view models.
- Added support for Keyed dependencies.
- Added a `MvvmObservableValidator`, similar to `DataAnnotationsValidator` but provides support for `ObservableValidator`. This means we listen to the `ErrorsChanged` on the `ObservableValidator` and pass those errors to the `EditContext`. The benefit is that the validation is handled in the `ObservableValidator`  and the view model and view have access to it, which opens the door for future enhancements like clearing errors on `ObservableValidator` and the `view` reacting to this, etc.
- Added `ParameterResolver` and `ViewParameterAttribute` to enable parameter resolution in `view models`.

## Other changes

- Added XML/Code documentation.
- Updated sample projects.
- Cleaned up some code.
- Renamed `ViewModels` folder to `ComponentModel`
- Add example for Keyed dependencies.
- Improve the Disposable pattern in the `ViewModelBase` & `MvvmComponentBase` classes
- Added examples for `MvvmOwningComponentBase`
- Added unit and component (using Bunit) tests.
- Update application version to 2.0.0
- Updated NuGet packages

## TODO:
- Update README to show how to use new features.

gragra33 commented:
@mishael-o As it stands, it looks ready for release. Are you able to clean up, clomplete comments, and add some tests?

mishael-o commented:
> @mishael-o As it stands, it looks ready for release. Are you able to clean up, complete comments, and add some tests?

Cool. I will add the tests, comments and do some clean up. Will push once done.

mishael-o commented:
@gragra33, the pull request is ready for review. I need to update the README with instructions on using the new features but feel free to review the rest in the meantime. Additionally, I've revised the description; please let me know if there's anything you're not comfortable with so we can discuss it.

I will try and get to the README updates this week.

gragra33 commented:
@mishael-o Totally agree bumping version to 2.0.0 + readme requires updating. Job well done! 

All looks good, except, I'm checking the code on a new clean machine. The WASM sample did not run unless using .Net v8.0.7. Did not run on version 8.0.6. Test run on both VS 2022 & Rider. Unclear on VS however Rider suggested updating the version.

I was thinking that we should leave the old methods and mark as depreciated pointing to the new versions however based on the changes, I think leaving them as breaking is the right choice. Maybe we need to add migration notes for these and the new view model auto registration and discovery using the new scope attribute.

Make sure that you give yourself credit for the work done.

mishael-o commented:
@gragra33 ok great, I will bump the version number.

> All looks good, except, I'm checking the code on a new clean machine. The WASM sample did not run unless using .Net v8.0.7. Did not run on version 8.0.6. Test run on both VS 2022 & Rider. Unclear on VS however Rider suggested updating the version.

Yeah, we are using a wildcard in the Nuget package reference. Hence, it automatically bumped it to the latest version (8.0.7), making the Nuget package dependent on that version or greater. 

![image](https://github.com/user-attachments/assets/45dbec68-e38e-4966-b52e-2618c394eb6c)

I will pin the version version number to 8.0.7 so we control the version updates.

> I was thinking that we should leave the old methods and mark as depreciated pointing to the new versions however based on the changes, I think leaving them as breaking is the right choice. Maybe we need to add migration notes for these and the new view model auto registration and discovery using the new scope attribute.

Yeah, I prefer deprecating the old methods and causing breaking changes (easier to manage). I will add migration notes to the README.

gragra33 commented:
> Yeah, we are using a wildcard in the Nuget package reference. Hence, it automatically bumped it to the latest version (8.0.7), making the Nuget package dependent on that version or greater.
> 
> ![image](https://private-user-images.githubusercontent.com/49244132/356891677-45dbec68-e38e-4966-b52e-2618c394eb6c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0MDYzODYsIm5iZiI6MTcyMzQwNjA4NiwicGF0aCI6Ii80OTI0NDEzMi8zNTY4OTE2NzctNDVkYmVjNjgtZTM4ZS00OTY2LWI1MmUtMjYxOGMzOTRlYjZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODExVDE5NTQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkYWNhNWJhNDMzN2YwMmQ3ZWVhYmQ1OWRjODRiZWM2ZGFiN2E5ZmI4OTUxZWI1MmZjNjk2MzFmZTY5NTU0NDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GRAd92C1WFjcX1br0dH1YfP2mp--7s8njiJeNHzaujE)
> 
> I will pin version number to 8.0.7 so we control the version updates.

Strangely, only the sample WASM version was affected. The server & web app versions ran with v8.0.6. Best to mention the version in the readme.

gragra33 commented:
@mishael-o I've been thinking about this for a while. Also, add `SetParametersAsync(ParameterView parameters)` method to `MvvmComponentBase` for passing to `ViewModelBase`. While manual processing is required, there is no need for double handling. A helper method could be added to use reflection to assign the properties with the `ParameterAttribute` on the ViewModel or done manually, no different to doing it in the Compenent itself.

mishael-o commented:
> @mishael-o I've been thinking about this for a while. Also, add `SetParametersAsync(ParameterView parameters)` method to `MvvmComponentBase` for passing to `ViewModelBase`. While manual processing is required, there is no need for double handling. A helper method could be added to use reflection to assign the properties with the `ParameterAttribute` on the ViewModel or done manually, no different to doing it in the Compenent itself.

Hey @gragra33, I see what you mean, then let's implement the `ParameterResolver` that I suggested in #6 , and then this can be done automatically without for extra logic on the `ViewModelBase`

gragra33 commented:
> Hey @gragra33, I see what you mean, then let's implement the `ParameterResolver` that I suggested in [#6](https://github.com/gragra33/Blazing.Mvvm/issues/6#issuecomment-1915536731), and then this can be done automatically without extra logic on the `ViewModelBase`

Sound good!

mishael-o commented:
> > Hey @gragra33, I see what you mean, then let's implement the `ParameterResolver` that I suggested in [#6](https://github.com/gragra33/Blazing.Mvvm/issues/6#issuecomment-1915536731), and then this can be done automatically without extra logic on the `ViewModelBase`
> 
> Sound good!

Do you want it as part of this release/PR, or we should plan for the next one?

gragra33 commented:
> Do you want it as part of this release/PR, or we should plan for the next one?

This release is overdue. It depends on how long it would take. I don't think that another week will make a difference.

mishael-o commented:
> > Do you want it as part of this release/PR, or we should plan for the next one?
> 
> This release is overdue. It depends on how long it would take. I don't think that another week will make a difference.

It should be a week or two depending on my schedule (being optimistic 😅). I will add it to this PR.

gragra33 commented:
> It should be a week or two depending on my schedule (being optimistic 😅). I will add it to this PR.

This will be a welcome addition. Maybe add an `IsHandled` to allow bypassing calling the base implementation in the `ComponentBase` class. This would be an enhancement to the Microsoft `SetParametersAsync` method, either as a param or return type.

mishael-o commented:
> > It should be a week or two depending on my schedule (being optimistic 😅). I will add it to this PR.
> 
> This will be a welcome addition. Maybe add an `IsHandled` to allow bypassing calling the base implementation in the `ComponentBase` class. This would be an enhancement to the Microsoft `SetParametersAsync` method implementation, either as a param or return type. Thoughts?

I see what you mean, but wouldn't the logic below in the `MvvmComponentBase` be sufficient? 

```csharp
public override async Task SetParametersAsync(ParameterView parameters)
{
    ParameterResolver.ResolveParameters(parameters, this, ViewModel);

    await base.SetParametersAsync(ParameterView.Empty);
    await ViewModel.SetParametersAsync();
}
```

Also, we will still call the base `SetParametersAsync` but we pass an empty parameter list as we will handle all the parameter resolutions.

https://learn.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?view=aspnetcore-8.0#when-parameters-are-set-setparametersasync
![image](https://github.com/user-attachments/assets/a7af1109-b133-4f8d-9fb8-b0874689a0c7)

gragra33 commented:
> I see what you mean, but wouldn't the logic below in the `MvvmComponentBase` be sufficient?

We can't assume how it is used. Someone could choose to use parameters in both the Component and the ViewModel. Let them choose ViewModel only.

mishael-o commented:
> > I see what you mean, but wouldn't the logic below in the `MvvmComponentBase` be sufficient?
> 
> We can't assume how it is used. Someone could choose to use parameters in both the Component and the ViewModel. Let them choose ViewModel only.

I see what you mean, I assumed that most users would want to use the parameters in the view model only. We can set the values on both the view and view model to allow flexibility. 

```csharp
public override async Task SetParametersAsync(ParameterView parameters)
{
    // Set parameters in the view model. Opt-in only for properties with attributes [Parameter] (or a custom one [ViewParameter]) are set.
    ParameterResolver.ResolveParameters(parameters, ViewModel);
    
    // Set parameters on the View, use the default implementation
    await base.SetParametersAsync(parameters);
    await ViewModel.SetParametersAsync();
}
```

gragra33 commented:
> I see what you mean, I assumed that most users would want to use the parameters in the view model only. We can set the values on both the view and view model to allow flexibility.

I was more thinking of:
```csharp
public override async Task SetParametersAsync(ParameterView parameters)
{
    // Set parameters in the view model. Opt-in only for properties with attributes [Parameter] (or a custom one [ViewParameter]) are set.
   if (! ParameterResolver.ResolveParameters(parameters, ViewModel)) // not flagged as handled
   { 
        // Set parameters on the View, use the default implementation
        await base.SetParametersAsync(parameters);
    }
    await ViewModel.SetParametersAsync();
}
```
This way we can avoid additional reflection by the base method.

mishael-o commented:
> This way we can avoid additional reflection by the base method.
> 
> Not sure what the overhead cost would be unless we benchmark it.

That's for happy paths, when all the parameters are used only in the view model, but like you said we are not sure where the parameters will be used. Consumers can have some parameters in the view model and others in the view.

Also, I am not so keen on passing `ParameterView` to the view model, feels like we are coupling some view elements to the view model. The `ParameterView` can be converted to an `IReadOnlyDictionary<string, object?>` with `parameters.ToDictionary()` then we could pass that to the view model. Regardless, I am not entirely opposed to passing the  `ParameterView` to the view model.

With this approach, we can avoid reflection on the view model side as there wouldn't be a need for `ParameterResolver` because the consumers would have to write custom logic to handle parameters. I can see the flexibility this allows but I can also see the drudgery of doing this for lots of view models.

For the `base.SetParametersAsync(parameters);` the types and property setters are cached, so reflection is done the first time to resolve that and then subsequent calls are handled by the cached data. https://source.dot.net/#Microsoft.AspNetCore.Components/Reflection/ComponentProperties.cs,69ab4ed7043d9cb1,references

I will benchmark these two approaches, and post the results here, then we can decide which we go for, or maybe a combination of both:

- Parameters fully resolved via reflection (ParameterResolver and `base.SetParametersAsync(parameters);`)
- Parameters resolved with custom code in the view model and the rest with `base.SetParametersAsync(parameters);` in the view.

gragra33 commented:
@mishael-o I looked at Microsoft's code too. Yes, I thought about `parameters.ToDictionary()`, as well, before posting here. Maybe I'm overthinking it and the impact may be minimal. Only benchmarks will tell us.

mishael-o commented:
Hey @gragra33, I did some benchmarking and these are the results.

### ParameterResolutionBenchmark

This benchmark compares the `manual` resolution and the `ParameterResolver` resolution to the `Baseline` which uses the `ParameterView.SetParameterProperties` provided by the SDK.

| Method                                      | Mean     | Error     | StdDev    | Median   | Ratio | RatioSD | Allocated | Alloc Ratio |
|-------------------------------------------- |---------:|----------:|----------:|---------:|------:|--------:|----------:|------------:|
| Baseline                                    | 3.559 us | 0.1003 us | 0.4954 us | 3.300 us |  1.02 |    0.18 |     400 B |        1.00 |
| 'Set parameters of VM manually'             | 3.177 us | 0.0444 us | 0.2142 us | 3.100 us |  0.91 |    0.12 |     400 B |        1.00 |
| 'Set parameters of VM with resolver'        | 6.697 us | 0.1840 us | 0.8932 us | 6.300 us |  1.91 |    0.34 |     496 B |        1.24 |
| 'Set parameters of cached VM with resolver' | 6.382 us | 0.1682 us | 0.8149 us | 6.100 us |  1.82 |    0.31 |     496 B |        1.24 |

The `ParameterResolver` is competitive with the `Baseline` and `manual` resolution with regards to memory, however, with execution time it's a bit slower.

 `ParameterResolver` and `Baseline` both use reflection under the hood, however, the reason why `Baseline` is faster is because of some clever code done in `SetParameterProperties` to avoid hashing when possible (as this incurs the most cost) when looking for the property setter for the parameter name.

### RenderWithParameterBenchmark

This benchmark uses `Bunit` to render an empty (no HTML, only parameters defined in the .razor file) component and pass parameters to it. It may not be the most accurate as it is mostly benchmarking the rendering done by `BUnit` but this helps to identify if parameter resolution affects rendering in any impactful way.

| Method                                                            | Mean     | Error    | StdDev   | Median   | Ratio | RatioSD | Allocated | Alloc Ratio |
|------------------------------------------------------------------ |---------:|---------:|---------:|---------:|------:|--------:|----------:|------------:|
| Baseline                                                          | 776.1 us | 30.68 us | 158.0 us | 750.6 us |  1.04 |    0.29 | 184.55 KB |        1.00 |
| 'Set parameters in VM (manually) & View (base)'                   | 791.2 us | 27.51 us | 140.7 us | 763.8 us |  1.06 |    0.27 | 190.74 KB |        1.03 |
| 'Set parameters in VM (manually) & View (SetParameterProperties)' | 803.2 us | 33.93 us | 173.2 us | 766.4 us |  1.07 |    0.31 | 190.74 KB |        1.03 |
| 'Set parameters in VM (manually) only'                            | 812.6 us | 31.13 us | 158.6 us | 787.7 us |  1.09 |    0.29 | 190.74 KB |        1.03 |
| 'Set parameters in VM (resolver) & View (base)'                   | 822.5 us | 33.06 us | 166.6 us | 814.0 us |  1.10 |    0.30 | 188.81 KB |        1.02 |
| 'Set parameters in VM (resolver) & View (SetParameterProperties)' | 802.5 us | 27.83 us | 141.8 us | 793.2 us |  1.07 |    0.28 | 189.14 KB |        1.02 |
| 'Set parameters in VM (resolver) only'                            | 739.4 us | 21.22 us | 108.1 us | 701.5 us |  0.99 |    0.23 | 189.14 KB |        1.02 |

### SetParameterAfterRenderingBenchmark

Similar to the benchmark above this uses `Bunit` however, we set parameters after the component has been rendered and perform a rerender. This reduces the noise and helps to identify if parameter resolution affects rendering in any impactful way.

| Method                                                            | Mean     | Error    | StdDev   | Median   | Ratio | RatioSD | Allocated | Alloc Ratio |
|------------------------------------------------------------------ |---------:|---------:|---------:|---------:|------:|--------:|----------:|------------:|
| Baseline                                                          | 50.44 us | 2.872 us | 14.69 us | 46.40 us |  1.08 |    0.42 |   8.88 KB |        1.00 |
| 'Set parameters in VM (manually) & View (base)'                   | 55.83 us | 2.886 us | 14.81 us | 53.10 us |  1.19 |    0.44 |      9 KB |        1.01 |
| 'Set parameters in VM (manually) & View (SetParameterProperties)' | 53.35 us | 2.662 us | 13.54 us | 49.65 us |  1.14 |    0.41 |      9 KB |        1.01 |
| 'Set parameters in VM (manually) only'                            | 50.52 us | 2.648 us | 13.49 us | 46.90 us |  1.08 |    0.40 |      9 KB |        1.01 |
| 'Set parameters in VM (resolver) & View (base)'                   | 56.26 us | 3.036 us | 15.44 us | 50.50 us |  1.20 |    0.45 |   9.09 KB |        1.02 |
| 'Set parameters in VM (resolver) & View (SetParameterProperties)' | 57.00 us | 2.668 us | 13.35 us | 54.50 us |  1.22 |    0.42 |   9.09 KB |        1.02 |
| 'Set parameters in VM (resolver) only'                            | 51.73 us | 2.513 us | 12.73 us | 48.30 us |  1.10 |    0.39 |   9.09 KB |        1.02 |

I have attached the project file, you can go through it and run the command below to validate the findings.

``` bash
dotnet run -c release
```

[BenchMarkBlazor.zip](https://github.com/user-attachments/files/16697935/BenchMarkBlazor.zip)

gragra33 commented:
@mishael-o impressed. The impact is minimal. Let's go with your suggestion.

mishael-o commented:
Hey @gragra33, I haven't had the time to work on this unfortunately, (life 😅 and other commitments), but I will target for this to be completed and merged by the end of October.

gragra33 commented:
> Hey @gragra33, I haven't had the time to work on this unfortunately, (life 😅 and other commitments), but I will target for this to be completed and merged by the end of October.

Yes, life can get in the way. I'm waiting to add .Net 9.0 support in prep for the release.

gragra33 commented:
@mishael-o I'm just checking in to see how you are tracking ... another PR is waiting on this.

mishael-o commented:
Hey @gragra33, yeah I did see. I will do some work over the weekend to complete the implementation for the resolver logic. 

Then it will just be docs and README left to update.

mishael-o commented:
@gragra33 changes in the latest push. 

Updated project versions to 2.0.0 and enhanced parameter resolution:
- Added `IParameterResolver` and `ParameterResolver` for setting parameters.
- Introduced `ViewParameterAttribute` for view model properties.
- Updated `MvvmComponentBase`, `MvvmLayoutComponentBase`, and `MvvmOwningComponentBase` to use `IParameterResolver`.
- Enhanced `AddMvvm` service registration with `ParameterResolutionMode`.
- Refactored tests and added new ones for parameter resolution.
- `Blazing.Mvvm.Sample.Wasm.csproj`: Updated to 8.0.10 for `Microsoft.AspNetCore.Components.WebAssembly`, `Microsoft.AspNetCore.Components.WebAssembly.DevServer`, and `Microsoft.AspNetCore.WebUtilities`.
- `Blazing.Mvvm.Sample.WebApp.Client.csproj`: Updated to 8.0.10 for `Microsoft.AspNetCore.Components.WebAssembly` and `Microsoft.AspNetCore.WebUtilities`.
- `Blazing.Mvvm.Sample.WebApp.csproj`: Updated to 8.0.10 for `Microsoft.AspNetCore.Components.WebAssembly.Server`.
- `Blazing.Mvvm.Tests.csproj`: Updated `bunit` to 1.33.3, `FluentAssertions` to 6.12.1, `Microsoft.NET.Test.Sdk` to 17.11.1, `xunit` to 2.9.2, `xunit.runner.visualstudio` to 2.8.2, and `coverlet.collector` to 6.0.2.
- `Blazing.Mvvm.csproj`: Updated `CommunityToolkit.Mvvm` to 8.3.2 and `Microsoft.AspNetCore.Components.Web` to 8.0.10.

You can begin testing and reviewing while I update the README file.

gragra33 commented:
> @gragra33 changes in the latest push.
> 
> Updated project versions to 2.0.0 and enhanced parameter resolution:
> 
> * Added `IParameterResolver` and `ParameterResolver` for setting parameters.
> * Introduced `ViewParameterAttribute` for view model properties.
> * Updated `MvvmComponentBase`, `MvvmLayoutComponentBase`, and `MvvmOwningComponentBase` to use `IParameterResolver`.
> * Enhanced `AddMvvm` service registration with `ParameterResolutionMode`.
> * Refactored tests and added new ones for parameter resolution.
> * `Blazing.Mvvm.Sample.Wasm.csproj`: Updated to 8.0.10 for `Microsoft.AspNetCore.Components.WebAssembly`, `Microsoft.AspNetCore.Components.WebAssembly.DevServer`, and `Microsoft.AspNetCore.WebUtilities`.
> * `Blazing.Mvvm.Sample.WebApp.Client.csproj`: Updated to 8.0.10 for `Microsoft.AspNetCore.Components.WebAssembly` and `Microsoft.AspNetCore.WebUtilities`.
> * `Blazing.Mvvm.Sample.WebApp.csproj`: Updated to 8.0.10 for `Microsoft.AspNetCore.Components.WebAssembly.Server`.
> * `Blazing.Mvvm.Tests.csproj`: Updated `bunit` to 1.33.3, `FluentAssertions` to 6.12.1, `Microsoft.NET.Test.Sdk` to 17.11.1, `xunit` to 2.9.2, `xunit.runner.visualstudio` to 2.8.2, and `coverlet.collector` to 6.0.2.
> * `Blazing.Mvvm.csproj`: Updated `CommunityToolkit.Mvvm` to 8.3.2 and `Microsoft.AspNetCore.Components.Web` to 8.0.10.
> 
> You can begin testing and reviewing while I update the README file.

I'll look at it this weekend. I need to update `MvvmNavigationManager` and `MvvmNavLink` to support keyed DI.

gragra33 commented:
@mishael-o under 3 weeks until 9.0 launch. How are you going?

mishael-o commented:
> @mishael-o under 3 weeks until 9.0 launch. How are you going?
> 
> `MvvmNavigationManager`, & `MvvmNavLink` updates are ready for keyed support + sample page. I'm just waiting for you to complete this PR...

Hey, @gragra33. Apologies, I've been busy, but I will push the docs changes by Wednesday or Thursday (I promised I would get this sorted before the end of October😅, so I'm still aiming for that), and then we can close this.

mishael-o commented:
@gragra33 I have pushed the document updates. You are free to update or revert any changes you are not comfortable with.

gragra33 commented:
Adding Auto Registration by @mishael-o
